### DISCUSSION TITLE
Lidar initialization: pitch, roll, yaw

### DISCUSSION THREAD

**Discussion by fernandodamasio** [microsoft/AirSim#3647] [Support Q&A]
Hi,

I'm coding an algorithm to build a HD Map of my environment.

I'm using `simGetGroundTruthKinematics `to return position and orientation (pitch, roll, yaw).

In the lidar.md file, it's written:

> - The API returns a Point-Cloud as a flat array of floats along with a timestamp of the capture.
> - The floats represent [x,y,z] coordinate for each point hit within the range in the last scan.
> - The coordinates are in the local vehicle NED like all other AirSim APIs.

My question is:

If I initialize the Lidar in settings.json defining x, y, z, pitch, roll, yaw with values, different then default values, the point cloud will be defined taking in consideration the rotation and position of the sensor or will return the coordinates in the local vehicle NED anyway?

Best regards,
Fernando Damasio

**Comment by bmalhigh** [comment]
Fernando, right now, the Point-Cloud that is returned has points in the vehicle frame.
Based on your use case, is that reasonable?

**Comment by fernandodamasio** [comment]
Hi @bmalhigh 

Right now it's reasonable to work with the lidar in the vehicle frame. TKS!

However, I have more questions about it.

I just wrote a code to create an HD Map of my environment. Next, you can find links to help understand the problem:

- Here you can see a video of the environment: https://www.youtube.com/watch?v=oDBrac6QwOs
- Lidar mapping code: https://github.com/skoods-org/welcome/blob/master/PythonExamples/lidar_mapping.py
- The project: https://github.com/skoods-org/welcome

1. I can see that the lidar returns the point cloud also taking into consideration the offset and rotation of the car. What I mean is, when driving the car and getting the points, the points will be updated based on the initial position of the car. Is this true? Because in real life, the points are always relative to the sensor position and we have to calculate the offset and rotation using other sensors, like the IMU.

2. What are the default values of the Lidar settings for the .json file?

3. I drew the track and guardrail of my environment using a mesh and a spline. But the lidar is not returning the points of this mesh. I know it because the guardrail is not in the point cloud. Do you know how to solve it? Maybe something related to collisions in Unreal? I'm including 2 images...

![image](https://user-images.githubusercontent.com/21197744/47678248-11855080-dba0-11e8-84d3-c60f371e0503.png)

![image](https://user-images.githubusercontent.com/21197744/47678273-2235c680-dba0-11e8-82d3-f9646a44a3d7.png)

4. Do you have a helper code to rotate and translate the points to become relative to the sensors?

I appreciate your help.

Best regards,
Fernando Damasio

**Comment by bmalhigh** [comment]
w.r.t. #2 -- Please see initializeFromSettings(...) method in https://github.com/Microsoft/AirSim/blob/master/AirLib/include/sensors/lidar/LidarSimpleParams.hpp
You will some car and multirotor specific defaults there.

w.r.t. #3 -- As you mentioned, the lidar ray-casts uses the collision channel; so only objects that have a collision-mesh will be hit. Look at the GetObstacle(...) method in https://github.com/Microsoft/AirSim/blob/master/Unreal/Plugins/AirSim/Source/AirBlueprintLib.h
Wondering if we need to play around with "ECollisionChannel collision_channel = ECC_Visibility" 

Let me get back to you on #1 and #4.

**Reply by vinayakap** [reply]
Hi @bmalhigh 

did you find anything on 4th one - Do you have a helper code to rotate and translate the points to become relative to the sensors?

**Comment by fernandodamasio** [comment]
Hi @bmalhigh 

2. TKS!

3. The spline mesh I'm using has a collision defined when drawing the spline, programmatically. It's free in the Unreal Launcher.

![image](https://user-images.githubusercontent.com/21197744/47749565-25968400-dc6c-11e8-94c6-8005ae68439d.png)

Do you see a way to work lidar with it? I don't have too much knowledge about creating meshes.

About 1. and 4. I will wait for your return.

TKS!!!

Regards,
Fernando Damasio

**Comment by TheGreatGalaxy** [comment]
Hi, I want to sure that which coordinate system is adopted to represent the coordinate of  each point in the points cloud. Is the play start's coordinate system? I find that the car's position captured by the "record" button seems not in accord  with the position captured by lidar.
